### PR TITLE
Add companies house search by name endpoint

### PR DESCRIPTION
**Pull request by richtier**
[this test](https://uktrade.atlassian.net/browse/ED-1345)

# successful lookup:

list

# validation error:

error-list

### CODE REVIEW

File: enrolment/views.py [hunk 1]
```diff
@@ -114,3 +115,22 @@ def get_context_data(self, *args, **kwargs):
...
+
+    def get(self, request, *args, **kwargs):
+        form = self.form_class(data=request.GET)
+        if not form.is_valid():
+            return JsonResponse(form.errors, status=400)
+        api_response = helpers.CompaniesHouseClient.search(
+            term=form.cleaned_data['term']
+        )
+        api_response.raise_for_status()
+        return JsonResponse(api_response.json()['items'], safe=False)
```
Comment (richtier) [inline enrolment/views.py]: we return all the data provided by CH. Realistically we only need `name` and `number`. I've returned all though to keep things simple as we don't really lose anything of value by being too verbose here.

File: ui/urls.py [hunk 2]
```diff
@@ -120,4 +121,10 @@
         require_get(company_proxies.CompanyPrivateAPIViewProxy.as_view()),
         name='api-external-company'
     ),
+    url(
+        r'^api/internal/companies-house-search/$',
```
Comment (richtier) [inline ui/urls.py]: this url will be requested dynamically by some javascript when the user clicks "search companies" on the landing page.

anyone could request this url, so we will need to add csrf protection. I'll add a ticket in the story for this.

File: enrolment/views.py (hunk 1 above)
Comment (konradko) [inline enrolment/views.py]: I think it is possible to specify the fields you want in the request (it is in the CH API docs) - we should be a good citizen and request only want is needed

File: ui/urls.py (hunk 2 above)
Comment (konradko) [inline ui/urls.py]: Very good point with the CSRF

File: enrolment/helpers.py [hunk 3]
```diff
@@ -63,22 +59,43 @@ def has_company(sso_user_id):
...
-        logger.error(MESSAGE_AUTH_FAILED)
-    return response
-
-
-def get_companies_house_profile(number):
-    url = COMPANY_PROFILE_URL.format(number=number)
-    return companies_house_client(url)
-
+class CompaniesHouseClient:
+    secret = settings.COMPANIES_HOUSE_API_KEY
```
Comment (konradko) [inline enrolment/helpers.py]: maybe `api_key` instead of `secret`?

File: enrolment/helpers.py [hunk 4]
```diff
@@ -63,22 +59,43 @@ def has_company(sso_user_id):
...
-    if response.status_code == http.client.UNAUTHORIZED:
-        logger.error(MESSAGE_AUTH_FAILED)
-    return response
-
-
-def get_companies_house_profile(number):
-    url = COMPANY_PROFILE_URL.format(number=number)
-    return companies_house_client(url)
-
+class CompaniesHouseClient:
```
Comment (konradko) [inline enrolment/helpers.py]: Nice client

File: enrolment/helpers.py [hunk 5]
```diff
@@ -63,22 +59,43 @@ def has_company(sso_user_id):
...
+
+    @classmethod
+    def retrieve_address(cls, number):
+        url = cls.endpoints['address'].format(number=number)
+        return cls.get(url)
+
+    @classmethod
+    def search(cls, term):
+        url = cls.endpoints['search']
+        return cls.get(url, {'q': term})
```
Comment (konradko) [inline enrolment/helpers.py]: we might perform validation on the term (same that CH does) and save a roundtrip in cases input is invalid

File: enrolment/views.py (hunk 1 above)
Comment (richtier) [inline enrolment/views.py]: I heard that too, but last time I checked the docs there was no way to do that. I'll check the docs again

File: enrolment/helpers.py (hunk 5 above)
Comment (richtier) [inline enrolment/helpers.py]: sure. I will add the validation to the form.

Comment (richtier) [inline enrolment/helpers.py]: it seems their API has no max limit on the search term.
```
CompaniesHouseClient.search('thing'*10000)
{'items_per_page': 20, 'total_results': 0, 'start_index': 0, 'page_number': 1, 'items': [], 'kind': 'search#companies'}
```

also, they allow searching by company number and text.

Aside from the existing "this field is required", there is no other validation to add.

File: enrolment/views.py (hunk 1 above)
Comment (richtier) [inline enrolment/views.py]: I checked the docs again, searched the forum, and googled - and it seems they do not support this feature.

**Pull request merged**